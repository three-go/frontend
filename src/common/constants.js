const iconNames = {
  heart: "heart",
  brokenHeart: "heart-broken-outline",
  leftArrow: "arrow-left-thick",
  flag: "flag",
};

const iconSizes = {
  headerChance: 28,
  headerLeftArrow: 40,
  headerFlag: 32,
};

const colors = {
  black: "#000000",
  dark: "#212529",
  blueGray: "#2F455C",
  ligthGray: "#5c6977",
  gray: "#96A1A8",
  ivory: "#FCF8F6",
  white: "#ffffff",
  red: "#f45a5a",
  darkRed: "#c92a2a",
  green: "#21D0B2",
  tealGreen: "#00BBD1",
  turquoise: "#00E0EA",
  lightBlue: "#5fceec",
};

const gameStatus = {
  ready: "ready",
  start: "start",
  input: "input",
  end: "end",
};

const navigations = {
  main: "Main",
};

const map = {
  width: 300,
  height: 450,
};

const facePosition = {
  top: 60,
  bottom: 390,
};

const faceYawAngleRange = {
  front: {
    low: -10,
    high: 10,
  },
  left: -36,
  right: 36,
};

export {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
};
