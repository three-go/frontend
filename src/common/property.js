import uuid from "react-native-uuid";

const gameNames = {
  game1: "GAME 1",
  game2: "GAME 2",
};

const gameColors = {
  game1: "#00BBD1",
  game2: "#00E0EA",
};

const game1Guide = [
  {
    title: "게임 설명",
    titleContent:
      "• 시작지점부터 종료지점까지 장애물을 피해 미로를 탈출하는 게임입니다.\n• 총 3개의 stage로 구성되어 있습니다.",
    children: [
      {
        id: uuid.v4(),
        title: "Map",
        content:
          "• 일정시간동안 맵을 기억하세요.\n• 탈출하는 최단거리를 기억하세요",
        image: require("../../public/assets/images/game1/map.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Face",
        content: "• 얼굴로 캐릭터가 움직이는 방향을 기록하세요.",
        image: require("../../public/assets/images/game1/camera.png"),
        icon: "face-recognition",
      },
      {
        id: uuid.v4(),
        title: "Play",
        content: "• 기록된 방향에 따라 캐릭터가 움직입니다.",
        image: require("../../public/assets/images/game1/play.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Next Stage",
        content: "• 캐릭터가 종료지점에 도착하면 다음 스테이지로 이동합니다.",
        image: require("../../public/assets/images/game1/next.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Fail",
        content:
          "• 캐릭터가 종료지점에 도착하지 않으면 목숨 하나가 소진된다.\n• 다시시도 또는 메인화면으로 돌아갈수 있다.",
        image: require("../../public/assets/images/game1/fail.jpg"),
      },
      {
        id: uuid.v4(),
        title: "End",
        content: "• 목숨을 모두 소진하면 게임이 종료된다.",
        image: require("../../public/assets/images/game1/end.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Success",
        content: "• 마지막 stage까지 성공하면 내 점수를 입력할 수 있습니다.",
        image: require("../../public/assets/images/game1/success.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Score",
        content: "• 지금까지 플래이한 점수를 볼 수 있습니다.",
        image: require("../../public/assets/images/game1/score.jpg"),
      },
    ],
  },
  {
    title: "플레이 방법",
    titleContent: "• 게임 진행 방법에 대하여 알려드리겠습니다",
    children: [
      {
        id: uuid.v4(),
        title: "Face",
        content:
          "• 위쪽: 얼굴의 눈 위치를 카메라 화면에 상단에 위치시키세요\n• 아래쪽: 얼굴의 눈 위치를 카메라 화면에 하단에 위치시키세요\n• 왼쪽: 얼굴의 고개를 왼쪽으로 돌리세요.\n• 오른쪽: 얼굴의 고개를 오른쪽으로 돌리세요.",
        image: require("../../public/assets/images/game1/camera.png"),
        icon: "face-recognition",
      },
      {
        id: uuid.v4(),
        title: "Play",
        content: "• 기록된 방향에따라 캐릭터가 자동으로 움직입니다.",
        image: require("../../public/assets/images/game1/play.jpg"),
      },
    ],
  },
];

const gameDescription = {
  game1: game1Guide,
  game2: [],
};

const gameKey = ["game1", "game2"];

const game = {
  keys: gameKey,
  names: gameNames,
  colors: gameColors,
  description: gameDescription,
};

const colors = {
  ivory: "#FCF8F6",
};

export { game, colors };
