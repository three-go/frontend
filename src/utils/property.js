const gameNames = {
  game1: "GAME 1",
  game2: "GAME 2",
};

const gameColors = {
  game1: "#00BBD1",
  game2: "#00E0EA",
};

const gameDescription = [["게임 설명 1111"], ["게임 설명 2222"]];

const gameKey = ["game1", "game2"];

const game = {
  keys: gameKey,
  names: gameNames,
  colors: gameColors,
  description: gameDescription,
};

export { game };
