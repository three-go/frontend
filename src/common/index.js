import {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
} from "./constants";
import { game } from "./property";

export {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
  game,
};
