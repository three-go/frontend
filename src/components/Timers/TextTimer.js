import React, { useEffect, useState } from "react";

import { View, Text, StyleSheet } from "react-native";
import PropTypes from "prop-types";

import { colors, textSizes, time } from "../../common/constants";
import { game } from "../../common/property";

const TextTimer = ({ onTimerEnd, status }) => {
  const [timerInfo, setTimerInfo] = useState(null);

  const handleTimerInfo = (status) => {
    switch (status) {
      case game.status.none:
        setTimerInfo({
          text: "게임시작",
          // count: time.readyTimerSet,
          count: 1,
          size: textSizes.medium,
        });
        break;
      case game.status.open:
        setTimerInfo({
          text: "맵이 가려지기",
          // count: time.startTimerSet,
          count: 1,
          size: textSizes.small,
        });
        break;
      case game.status.directionInput:
        setTimerInfo({
          text: "입력이 종료되기",
          // count: time.inputTimerSet,
          count: 1,
          size: textSizes.small,
        });
    }
  };

  const handleTimer = () => {
    if (!timerInfo) return;

    let timeoutId;

    if (timerInfo.count > time.zero) {
      timeoutId = setTimeout(() => {
        setTimerInfo((state) => {
          return {
            ...state,
            count: state.count - 1,
          };
        });
      }, time.textTimerInterval);
    }

    if (timerInfo.count === time.zero) {
      onTimerEnd();
    }

    return timeoutId;
  };

  useEffect(() => {
    handleTimerInfo(status);
  }, [status]);

  useEffect(() => {
    const timeoutId = handleTimer();

    return () => {
      clearTimeout(timeoutId);
    };
  }, [timerInfo, onTimerEnd, status]);

  return (
    <View style={styles.container}>
      <Text style={styles.text(timerInfo?.size)}>
        {timerInfo?.text}
        <Text style={styles.number(timerInfo?.size)}>{timerInfo?.count}</Text>초
        전
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    marginBottom: 8,
    alignItems: "center",
  },
  text: (size) => {
    return {
      color: colors.ivory,
      fontSize: size ? size : 10,
    };
  },
  number: (size) => {
    return {
      color: colors.ivory,
      fontSize: size ? size * 2 : 16,
    };
  },
});

export default React.memo(TextTimer);

TextTimer.propTypes = {
  onTimerEnd: PropTypes.func,
  status: PropTypes.string,
};
