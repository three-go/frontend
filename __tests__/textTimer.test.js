import React from "react";

import { render, act } from "@testing-library/react-native";

import "@testing-library/jest-native";
import "react-native-gesture-handler/jestSetup";
import { game } from "../src/common/property";
import TextTimer from "../src/components/Timers/TextTimer";

jest.mock("react-native-sound", () => {
  const Sound = () => {
    const obj = {};
    obj.sound = "music";
    return obj;
  };

  return Sound;
});

jest.useFakeTimers();

describe("TextTimer Test", () => {
  it("TextTimer Test", () => {
    const props = {
      onTimerEnd: jest.fn(),
      status: game.status.none,
    };

    const textTimer = render(<TextTimer {...props} />);
    const textTimerJson = textTimer.toJSON();

    expect(textTimerJson).toBeTruthy();
    expect(textTimerJson).toMatchSnapshot();
    expect(textTimer.getByTestId("timer")).toHaveTextContent("게임시작");
    expect(textTimer.getByTestId("timer")).toHaveTextContent("3초 전");

    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(textTimer.getByTestId("timer")).toHaveTextContent("2초 전");

    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(textTimer.getByTestId("timer")).toHaveTextContent("1초 전");
  });
});
