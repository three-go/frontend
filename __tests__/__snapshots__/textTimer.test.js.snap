// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextTimer Test TextTimer Test 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "marginBottom": 8,
    }
  }
  testID="timer"
>
  <Text
    style={
      Object {
        "color": "#FCF8F6",
        "fontSize": 24,
      }
    }
  >
    게임시작
    <Text
      style={
        Object {
          "color": "#FCF8F6",
          "fontSize": 48,
        }
      }
    >
      3
    </Text>
    초 전
  </Text>
</View>
`;
